File: gtci/03-sliding-window/cpp/5-longest-substring-with-same-letters-after-replacement.cpp
#include <algorithm>
#include <cassert>
#include <iostream>
#include <string>
#include <unordered_map>

using namespace std;

class Solution {
   public:
    int findLength(const string &str, int k) {
        int maxLength = 0;

        int windowStart = 0;
        int maxLetterFreq = 0;
        std::unordered_map<char, int> chars;
        for (int windowEnd = 0; windowEnd < str.size(); windowEnd++) {
            int windowSize = windowEnd - windowStart + 1;
            chars[str[windowEnd]] += 1;

            maxLetterFreq = std::max(maxLetterFreq, chars[str[windowEnd]]);

            if (windowSize - maxLetterFreq <= k)
                maxLength = std::max(maxLength, windowSize);

            while (windowSize - maxLetterFreq > k) {
                chars[str[windowStart]] -= 1;
                if (chars[str[windowStart]] == 0) chars.erase(str[windowStart]);

                windowStart++;
                windowSize = windowEnd - windowStart + 1;
            }
        }

        return maxLength;
    }
};

int main() {
    Solution sol;

    assert(sol.findLength("aabccbb", 2) == 5);
    assert(sol.findLength("abbcb", 1) == 4);
    assert(sol.findLength("abccde", 1) == 3);

    std::cout << "All test cases passed." << std::endl;
    return 0;
}

File: gtci/03-sliding-window/python3/5-longest-substring-with-same-letters-after-replacement.py
class Solution:
    def findLength(self, str1, k):
        max_length = 0

        window_start = 0
        majority_letter_freq = 0
        chars = {}
        for window_end in range(len(str1)):
            window_size = window_end - window_start + 1

            if str1[window_end] not in chars:
                chars[str1[window_end]] = 0
            chars[str1[window_end]] += 1

            majority_letter_freq = max(majority_letter_freq, chars[str1[window_end]])

            if window_size - majority_letter_freq <= k:
                max_length = max(max_length, window_size)

            while window_size - majority_letter_freq > k:
                chars[str1[window_start]] -= 1
                if chars[str1[window_start]] == 0:
                    del chars[str1[window_start]]

                window_start += 1
                window_size = window_end - window_start + 1

        return max_length


if __name__ == "__main__":
    sol = Solution()
    assert sol.findLength("aabccbb", 2) == 5
    assert sol.findLength("abbcb", 1) == 4
    assert sol.findLength("abccde", 1) == 3

    print("All test cases passed.")
